### PR TITLE
Update checkpointing directory

### PR DESCRIPTION
**Pull request by felipemello1**
Co-authored-by: vancoyendall <vancoykendall@gmail.com>#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [x] fix a bug
- [x] update tests and/or documentation
- [x] other (please add here)

TLDR using vLLM and Hugginface .frompretrained

1) train your model using torchtune main (nightlies). It will produce a folder like this:
image

2. Copy the contents of your latest epoch to base_model folder, which contains the checkpoint_dir original's content, without the model files (.pt, .bin, .safetensors):
```
cp /tmp/llama_3_2_1b/lora_single_device/epoch_2/* /tmp/llama_3_2_1b/lora_single_device/base_model
```

Making it look like this:
image

3. Now, you can use it with vLLM and Huggingface. There is one catch here: when using lora, we output MERGED weights AND the adapter. You should NOT use both at the same time. Either use ONLY the merged weights OR the BASE UNTRAINED MODEL + adapter.

3.1. Using hugginface .from_pretrained with BASE UNTRAINED MODEL + adapter

```
from peft import PeftModel
from transformers import AutoModelForCausalLM, AutoTokenizer

# Define the model and adapter paths
original_model_name = "meta-llama/Llama-3.2-1B-Instruct"
trained_model_path = "/tmp/torchtune/llama3_2_1B/lora_single_device/base_model"

model = AutoModelForCausalLM.from_pretrained(original_model_name)

# huggingface will look for adapter_model.safetensors and adapter_config.json
peft_model = PeftModel.from_pretrained(model, trained_model_path)

# Load the tokenizer
tokenizer = AutoTokenizer.from_pretrained(original_model_name)

# Function to generate text
def generate_text(model, tokenizer, prompt, max_length=50):
    inputs = tokenizer(prompt, return_tensors="pt")
    outputs = model.generate(**inputs, max_length=max_length)
    return tokenizer.decode(outputs[0], skip_special_tokens=True)

prompt = "Complete the sentence: 'Once upon a time...'"
print("Base model output:", generate_text(peft_model, tokenizer, prompt))
```

3.2. using the MERGED TRAINED MODEL with vLLM

IMPORTANT: this will not work right away. Your output directory has 2 files:
- ft-model-00001-of-00001.safetensors
- adapter_model.safetensors

vLLM doesnt know what is a model and what is an adapter. When it tries to load the adapter, it will raise an error. Therefore, delete adapter_model.safetensors from the folder, and it will work

```
rm /tmp/torchtune/llama3_2_1B/lora_single_device/base_model/adapter_model.safetensors
```

Now you can run vLLM locally
```
from vllm import LLM, SamplingParams

def print_outputs(outputs):
    for output in outputs:
        prompt = output.prompt
        generated_text = output.outputs[0].text
        print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
    print("-" * 80)

llm = LLM(
    model="/tmp/torchtune/llama3_2_1B/lora_single_device",
    load_format="safetensors",
    kv_cache_dtype="auto",
)
sampling_params = SamplingParams(max_tokens=16, temperature=0.5)

conversation = [
    {"role": "system", "content": "You are a helpful assistant"},
    {"role": "user", "content": "Hello"},
    {"role": "assistant", "content": "Hello! How can I assist you today?"},
    {
        "role": "user",
        "content": "Write an essay about the importance of higher education.",
    },
]
outputs = llm.chat(conversation, sampling_params=sampling_params, use_tqdm=False)
print_outputs(outputs)
```

#### Context

In torchtune's current state, if checkpoint_dir != outputdir, it breaks. Since the files are all mixed, saved as .pt and without the proper configs, its hard for users to readily use it with vllm/huggingface, resulting in issues such as https://github.com/pytorch/torchtune/issues/2048, https://github.com/pytorch/torchtune/issues/2025 and https://github.com/pytorch/torchtune/issues/2118.

This PR is **NOT** a major refactor. Everything is backwards compatible. The intention here is just to organize the output_dir and allow users to quickly use their models with HF and vLLM.

#### Changelog

1. The folder is automatically created/populated like described above

2. Initially, base_model has all the files from the checkpoint_dir, **except** those that end in .pt, .safetensors, .bin, etc

3. We save the original_model info in the adapter.config

4. naming got standardized

5. safetensors is the default for HF ckpt

6. Solved bugs when input_dir != output_dir

#### Next steps
Update docs

#### Unresolved issues
Left TODOs in the code, to be addressed in follow up PRs. It makes the code ugly, but we are due a refactor.

#### Test plan
- unit tests
- resumed pretrained
image
- ran vllm and hugginface

### CODE REVIEW

File: torchtune/_cli/download.py
Comment (felipemello1) [inline torchtune/_cli/download.py:99]: deprecated. Its a no-op.

File: torchtune/_cli/download.py
Comment (felipemello1) [inline torchtune/_cli/download.py:218]: related: https://github.com/pytorch/torchtune/pull/2026

File: torchtune/training/checkpointing/_utils.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_utils.py:229]: replaced it with "copy_files", so we save every file, and not only config

File: torchtune/training/checkpointing/__init__.py
Comment (joecummings) [inline torchtune/training/checkpointing/__init__.py:58]: I don't think these need to be exposed as public APIs right?

File: torchtune/_cli/download.py
Comment (felipemello1) [inline torchtune/_cli/download.py:179]: hf download

File: torchtune/_cli/download.py
Comment (felipemello1) [reply torchtune/_cli/download.py:217]: kaggle donwload

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:271]: not sure if i need every mkdir here. Will revise.

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:365]: are we ok with changing the default? I think it has some performance benefits when working with hf

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:360]: moved down

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:472]: i don't really like this. The other options are 1) to actually fix the issue, which is knowing save_adapter_only, or 2) silently let it happen, which is dangerous if the adapter + model was trained (e.g. embeddings), and the user forgot to change the file names.

Not sure if there is a 3rd

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:750]: no more .pt. Lets do .bin.

File: torchtune/training/checkpointing/__init__.py
Comment (felipemello1) [reply torchtune/training/checkpointing/__init__.py:58]: agreed! I was following the pattern, but i guess that these other constants are used in multiple files. I will revisit this.

File: recipes/configs/llama3_2/1B_lora_single_device.yaml
Comment (joecummings) [inline recipes/configs/llama3_2/1B_lora_single_device.yaml:39]: typo

File: torchtune/models/convert_weights.py
Comment (joecummings) [inline torchtune/models/convert_weights.py:233]: @ebsmothers  How was the PEFT model figuring this out before?

File: torchtune/training/__init__.py
Comment (joecummings) [inline torchtune/training/__init__.py:94]: Yeah these can live in individual files or a `constants` file within the training, but let's not expose these please.

File: torchtune/training/checkpointing/_checkpointer.py
Comment (joecummings) [inline torchtune/training/checkpointing/_checkpointer.py:173]:
```suggestion
           self._output_dir / BASE_MODEL_DIRNAME,
```

File: torchtune/training/checkpointing/_utils.py
Comment (joecummings) [inline torchtune/training/checkpointing/_utils.py:326]: Can you make ignore suffixes kwarg only?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (joecummings) [inline torchtune/training/checkpointing/_checkpointer.py:396]: Again I think this can just be joined with a forward-slash ?

File: torchtune/training/checkpointing/_utils.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_utils.py:326]: I don't follow. Can you give an example?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (joecummings) [reply torchtune/training/checkpointing/_checkpointer.py]: Easier to read IMO and I think we use it in other places, but es okay.

File: torchtune/training/checkpointing/_utils.py
Comment (joecummings) [reply torchtune/training/checkpointing/_utils.py:326]:
```python
def copy_files(
	input,
	output,
	*,
	ignore_suffixes
)
```

File: torchtune/_cli/download.py
Comment (SalmanMohammadi) [reply torchtune/_cli/download.py:99]: Should we add a deprecation warning rather than removing?

Comment (felipemello1) [reply torchtune/_cli/download.py:99]: good question! There is already a deprecation warning on huggingface side, so i think it should be ok to remove. Let me know if you disagree.

File: torchtune/models/convert_weights.py
Comment (ebsmothers) [reply torchtune/models/convert_weights.py:233]: Oh it wasn't. That's why we loaded PEFT models in two steps: 

```
model = AutoModelForCausalLM.from_pretrained(model_id)
peft_model = PeftModel.from_pretrained(model, checkpoint_dir)
```

instead of 

```
AutoModelForCausalLM.from_pretrained(checkpoint_dir)
```

I had a hacky version of this in #2026 but it was pointed out by @pbontrager that this shouldn't be present for models that are gonna get pushed to the hub (in that case we would want the hub model ID here, not a local path). I'm curious how we wanna address that here, should we wrap hugging-face CLI upload with something that will modify the path?

File: recipes/configs/llama3_2/1B_full_single_device.yaml
Comment (ebsmothers) [inline recipes/configs/llama3_2/1B_full_single_device.yaml:22]: What's the purpose of these config changes? Aren't these just a subset of the broader config changes in #2107?

File: torchtune/training/checkpointing/_utils.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_utils.py:247]: This is an arg to `Module.load_state_dict`, right? Not `torch.load`

File: torchtune/training/checkpointing/_utils.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_utils.py:326]: Call me crazy but do we need a full Python function reinventing recursive copy here? E.g. 

```
os.system(
f"rsync -av --ignore-existing {" ".join([f"--exclude *{}" for ignore_suffix in ignore_suffixes])} {input_dir} {output_dir}"
)
```

File: torchtune/training/checkpointing/_utils.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_utils.py:430]: add return types to these (e.g. here we return optional path but for recipe checkpoint we always return path)

File: torchtune/training/checkpointing/_utils.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_utils.py:398]: this can also return none, right?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_checkpointer.py:174]: Do we think that a single global variable for `SUFFIXES_TO_NOT_COPY` will suffice here? Maybe I'm paranoid but I anticipate this will be something people will need to hack into and change

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_checkpointer.py:403]: 😍

File: torchtune/training/checkpointing/_utils.py
Comment (SalmanMohammadi) [reply torchtune/training/checkpointing/_utils.py:326]: I just use [shutil.copy_tree](https://docs.python.org/3/library/shutil.html#shutil.copytree)

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_checkpointer.py:1057]: Is this todo applicable? Like what would we do with these configs for a Meta-format checkpoint anyways? (Maybe I am missing the point here)

File: recipes/configs/llama3_2/1B_full_single_device.yaml
Comment (felipemello1) [reply recipes/configs/llama3_2/1B_full_single_device.yaml:22]: i changed only these 2 configs here for testing. Then i created the other PR so its still easy to review this one. Once the other PR is merged, i will merge it into this one and retest it

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [inline torchtune/training/checkpointing/_checkpointer.py:436]: Thanks for enumerating the different cases here, but this is fairly convoluted. Why can't this be offloaded to a standalone utility?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py:174]: i should add to this function to not copy anything > x MiB, e.g. 200MiB. But i dont think that this should require user input. They have easily cp/rm files. Do you see an easy fix that doesnt greatly increase complexity?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_checkpointer.py:477]: I'm not sure I fully follow this, but for (1) couldn't it be done by just e.g. saving `save_adapter_weights_only` as part of the recipe state and pulling it from there?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py:1057]: it is just an useful map of ckpt file to weights. I guess no use in our current pipeline, but i think it makes sense to have all ckpts behave the same way, and the difference between them is mostly the model format, and **not** the metadata, naming, files, etc.

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py:436]: are you saying that the code can stay as it is, but have it as an utility? Or would you refactor the code itself?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py:477]: are you ok with saving it as part of the recipe_state? that would work.

>I'm not sure I fully follow this,
1. User trains a model that has both lora + finetuning (e.g. vision model)
2. User resumes from ckpt, and forgets to update the ckpt files
3. Since we are looking at the ckpt_dir, we will get the **untrained model**, which is a silent bug.

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_checkpointer.py:436]: I am saying why not just put all of this (i.e. L436-481) in a utility like `def infer_checkpoint_paths(self, ...)`? Then you can put all this case 1, case 2, etc into a nice clear docstring in a single place. We could think about refactoring too.. the fact that we are calling `_validate_hf_checkpoint_files` like ten different ways is not my favorite

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_checkpointer.py:174]: Yeah I agree with not increasing complexity here, I think it's OK to leave it like this for now

File: torchtune/models/convert_weights.py
Comment (felipemello1) [reply torchtune/models/convert_weights.py:233]: > peft_model = PeftModel.from_pretrained(model, checkpoint_dir)

I don't think that this works either. The info has to be inside of the adapter_config. Maybe if input_dir == output_dir?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py]: done

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py]: keeping it for safety

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py]: removed this. when lora=True, it will always get it from ckpt_dir (which is our current state anyway). Need a follow up PR to address it.

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py:294]: we do

File: tests/recipes/test_full_finetune_distributed.py
Comment (felipemello1) [inline tests/recipes/test_full_finetune_distributed.py:183]: test changes are related to finding the files. Before they were hardcoded.

Now we retrieve the epoch folder, get the suffix based on ckpt_type, and create the ckpt_name based on the defined standard.

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:411]: we save everything instead of just the config

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [inline torchtune/training/checkpointing/_checkpointer.py:385]: moved all this logic to self._checkpoint_paths = get_model_checkpoint_path

File: torchtune/training/checkpointing/_checkpointer.py
Comment (joecummings) [inline torchtune/training/checkpointing/_checkpointer.py:151]: ?

File: torchtune/training/checkpointing/_utils.py
Comment (joecummings) [reply torchtune/training/checkpointing/_utils.py:328]: Yeah wait actually is there a reason why can't use copytree? I'd rather that be maintained by core Python.

File: torchtune/training/checkpointing/_utils.py
Comment (joecummings) [inline torchtune/training/checkpointing/_utils.py:559]: I'm confused - you set input_dir with all the above switch statements and then you don't use it here?

Comment (felipemello1) [reply torchtune/training/checkpointing/_utils.py:559]: -.-`

File: torchtune/training/checkpointing/_checkpointer.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_checkpointer.py:151]: changed to "# TODO: support loading more than one file"

File: torchtune/training/checkpointing/_utils.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_utils.py:559]: i will update a couple of unit tests to use different output dir

File: torchtune/training/checkpointing/_utils.py
Comment (felipemello1) [reply torchtune/training/checkpointing/_utils.py:328]: beyond the suffixes, ii will also have to ignore .cache and .git__

idk, the function gives us some flexibility and its readable. But it seems to be 3x1. 

Can we leave it for the ckpt refactoring?

File: torchtune/training/checkpointing/_checkpointer.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_checkpointer.py]: Seems like there are a few follow-ups on this PR.. can we file an issue so we can track the different todos in a single place?

File: torchtune/training/checkpointing/_utils.py
Comment (ebsmothers) [reply torchtune/training/checkpointing/_utils.py:328]: Hmm I do like copy_tree and feel like it should be workable. But won't block on it, just include it in the follow-up task (as I mentioned above)

File: tests/recipes/test_full_finetune_single_device.py
Comment (ebsmothers) [inline tests/recipes/test_full_finetune_single_device.py:211]: remove

**Review by ebsmothers — APPROVED** [review]
Thanks for getting all of this sorted out! I know this was not an easy one. Once CI is green I think we're good to merge here. One request: can you update the summary to lead with the example usage within vLLM and PEFT? In case anyone is coming to this PR later I think that's what they will want to see. We should also add a section in the readme giving this process explicitly for better visibility.

**Comment by nouranali** [comment]
How would this method affect GPU consumption, having to load two models? While using only 1?

**Pull request merged**
